### PR TITLE
Reindex into new index when aliased

### PR DESCRIPTION
**Pull request by nickstenning**
Adds support for reindexing into a brand-new index when the current index is aliased.

Dropping support for unaliased indices will come later.

### CODE REVIEW

File: src/memex/search/index.py [hunk 1]
```diff
@@ -79,11 +82,35 @@ def delete(es, annotation_id):
...
-    indexing.index_all()
+    """Reindex all annotations into a new index, and update the alias."""
+
+    aliased = es.get_aliased_index() is not None
+
+    # If the index we're using is an alias, then we index into a brand new
+    # index and update the alias at the end.
+    if aliased:
+        new_index = es.index + '-' + _random_id()
+        configure_index(es, index=new_index)
```
Comment (chdorner) [inline src/memex/search/index.py]: I believe this will create an index with `number_of_replicas` set to 1 by default, which in our case (current) situation will make the cluster turn yellow because it can't assign the replica shards. Or am I missing something? (I'm also not really sure how to fix this)

File: src/memex/search/index.py [hunk 2]
```diff
@@ -92,25 +119,16 @@ class BatchIndexer(object):
...
-    def __init__(self, session, es_client, request):
+    def __init__(self, session, es_client, request, target_index=None):
         self.session = session
         self.es_client = es_client
         self.request = request
 
         # By default, index into the open index
-        self._target_index = self.es_client.index
-
-    def index_all(self):
```
Comment (nickstenning) [inline src/memex/search/index.py]: Simply that it didn't really seem like there was any reason that the `BatchIndexer` needed to know about anything other than where it was indexing to.

File: src/memex/search/index.py (hunk 1 above)
Comment (nickstenning) [inline src/memex/search/index.py]: We can sort this with an [index template](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html) in production.

File: src/memex/search/index.py (hunk 2 above)
Comment (chdorner) [inline src/memex/search/index.py]: I see this retrying functionality more part of a batch indexer, mainly because of the "batch" part of it where I don't want to care about transient errors of individual items in the batch

File: tests/memex/search/index_test.py [hunk 3]
```diff
@@ -93,43 +93,100 @@ def log(self, patch):
         return patch('memex.search.index.log')
 
 
-@pytest.mark.usefixtures('BatchIndexer', 'BatchDeleter')
+@pytest.mark.usefixtures('BatchDeleter', 'BatchIndexer', 'configure_index')
 class TestReindex(object):
-    def test_it_indexes_all_annotations(self, BatchIndexer):
-        index.reindex(mock.sentinel.session, mock.sentinel.es, mock.sentinel.request)
+    def test_indexes_annotations(self, es, indexer):
```
Comment (chdorner) [inline tests/memex/search/index_test.py]: I don't really see a reason to change the function name from "test_it_..." to "test_...", but that would be nicer in a separate commit for easier reviewing.

File: src/memex/search/index.py (hunk 2 above)
Comment (nickstenning) [inline src/memex/search/index.py]: Agreed, but the actual functionality of `BatchIndexer` is currently split between a) code to return a stream of annotations as fast as possible from a query, and b) code to manage indexing (including, possibly, retries). I'll try and clean this up in subsequent PRs.
